README.md
<p align="left">
    <img src="./docs/logo/php-starter.svg" alt="logo-text" width="200" />
    <br/>
    <h1>PHP Starter</h1>
    <p style="font-size: 1.2rem;">A quick start kit for PHP based website development with an optimised Gulp build process.</p>
</p>

<!-- Badges -->
<p align="left">
  <a href="https://travis-ci.org/brionmario/php-starter">
    <img src="https://travis-ci.org/brionmario/php-starter.svg?branch=master" alt="Build Status" height="18">
  </a>
  <a href="#contributors">
    <img src="https://img.shields.io/badge/all_contributors-1-orange.svg" alt="All Contributors" height="18">
  </a>
  <a href="https://david-dm.org/brionmario/php-starter">
    <img src="https://david-dm.org/brionmario/php-starter/status.svg" alt="dependencies Status" height="18">
  </a>
  <a href="https://david-dm.org/brionmario/php-starter?type=dev">
    <img src="https://david-dm.org/brionmario/php-starter/dev-status.svg" alt="devDependencies Status" height="18">
   </a>
  <a href="LICENSE.md">
    <img src="https://img.shields.io/badge/License-MIT-blue.svg" alt="License: MIT" height="18">
  </a>
  <a href="https://codecov.io/gh/brionmario/php-starter">
  <img src="https://codecov.io/gh/brionmario/php-starter/branch/master/graph/badge.svg" />
</a>
</p>

This repository could be used by anyone who's looking for a quick starter kit for PHP based website development. This boilerplate uses [Gulp](https://gulpjs.com/) for building and automating the workflow. Following are some key features we've included for you.

- Supports minification and concatination for production builds.
- Usage of [Sass: Syntactically Awesome Style Sheets](https://sass-lang.com/) over CSS.
- Ready to go build system using Gulp.
- Simplified development process with npm scripts.
- Easilily integrate your favourite php libraries using [Composer](https://getcomposer.org/).
- Ability to use your favorite bower packages.
- Organized and easy to use folder structure for beginners.

# Quick Links

| [Demo](https://php-starter-seed.herokuapp.com/) |
| ----------------------------------------------- |


# Quick start

> The generated project have dependencies that require `Node.js` together with `npm` & `Composer`.

**Install [Composer](https://getcomposer.org/) locally or globally by following the instructions listed [here](https://getcomposer.org/doc/00-intro.md#installation-linux-unix-macos).**

**Make sure you have [Node](https://nodejs.org/en/download/) version >= 8.0 and [npm](https://www.npmjs.com/) >= 5 or [Yarn](https://yarnpkg.com).**

```bash
# clone our repository
# --depth 1 removes all but one .git commit history
git clone --depth 1 https://github.com/brionmario/php-starter.git

# change the directory
cd php-starter

# install the npm dependencies listed on package.json
npm install

# install the composer dependencies listed on composer.json
Composer install

# start the development server
npm start

```

Once the dev server is fired up, it'll automatically open up a new tab. If not, manually navigate to the local URL listed on the terminal(usually [http://localhost:8000](http://localhost:8000)).

# Table of Contents

- [File Structure](#file-structure)
- [Getting Started](#getting-started)
  - [Prerequisites](#prerequisites)
  - [Setting Up](#setting-up)
  - [Running the app](#running-the-app)
- [Configuration](#configuration)
  - [Add third-party dependencies](#add-third-party-dependencies)
    - [Composer](#composer)
    - [Bower](#bower)
- [Styling](#styling)
- [Deployment](#deployment)
  - [Heroku](#deploy-to-heroku)
- [Built With](#built-with)
- [Contributing](#contributing)
- [Contributors](#contributors)
- [License](#license)

# File Structure

```
php-starter/
 ├── coverage/                      * unit test coverage reports
 ├── dev/                           * stores development build artifacts
 ├── dist                           * stores production build artifacts
 ├── docs/                          * contains documents and document resources
 ├── node_modules/                  * contains dependencies pulled from npm
 ├── src/                           * source folder
 │   ├── assets/                    * static assets such as images, icons, fonts goes here
 │   ├── components/                * reusable components can be placed here
 │   ├── scss/                      * styles folder
 │   │   ├── partials/              * place all the sass partial stylesheets in this folder
 │   │   └── styles.scss            * the main stylesheet for the project which gets compiled to CSS
 │   ├── scripts/                   * custom javascript script files
 │   ├── bower_components/          * third party bower libraries will be copied here
 │   ├── composer-demo.php          * demo of the composer dependency management
 │   ├── demo.php                   * file containing the demo sorce code
 │   └── index.php                  * entry php file
 ├── vendor/                        * packages pulled from composer will be stored here
 ├── .all-contributorsrc            * contains info ablout repo contributors
 ├── .babelrc                       * babel build config file
 ├── .bowerrc                       * bower config file
 ├── .buildpacks                    * contains heroku buildpacks
 ├── .dockerignore                  * contains files that are ignored from docker
 ├── .editorconfig                  * helps define and maintain consistent coding styles between different editors and IDEs
 ├── .eslintrc                      * ecmascript linting configuration file
 ├── .gitignore                     * contains files that are ignored from git
 ├── .npmrc                         * npm config file to house project wide custom configs
 ├── .nvmrc                         * node version manager config file
 ├── .pullapprove.yml               * pullapprove config file
 ├── .sass-lint.yml                 * sass linting configuration file
 ├── .travis.yml                    * travis ci configuration file
 ├── app.json                       * heroku manifest format for describing app
 ├── bower.json                     * contains bower dependencies
 ├── CHANGELOG.md                   * version changelog
 ├── composer.json                  * contains php project requirements
 ├── CONTRIBUTING.md                * project contributing guidelines
 ├── docker-compose.yml             * docker compose file
 ├── Dockerfile                     * docker config file
 ├── gulpfile.babel.js              * main buld configuration file. contains all the gulp tasks.
 ├── LICENSE.md                     * licensing information
 ├── package.json                   * contains all the npm scripts for building, running, deploying etc. and contains all the dependencies
 ├── Procfile                       * contains commands to be executed by heroku at app startup
 └── README.md                      * Readme file for the repository

```

# Getting Started

## Prerequisites

What you need to run this app:

- The generated project have dependencies that require `Node.js` together with `npm` and `Composer`.

> Make sure you have `Node.js`, `npm` & `Composer` installed by running the following simple commands on the command line and see if you're running the latest stable versions.

- Node.js - Type `node -v` on the terminal.
- npm - Type `npm -v` on the terminal.
- Composer - Type `Composer -v` on the terminal.

If you do not have them installed, click [here](https://nodejs.org/en/download/) and grab the latest stable version of `Node.js` and `npm` will be automatically installed along with it. Or if you have `brew` already installed in your local machine, execute `brew install node` command to get `node`.

To download composer, click [here](https://getcomposer.org/doc/00-intro.md#installation-linux-unix-macos) and follow the download instructions.

The project requires `bower` to install project scope dependencies such as `bootstrap`, `jquery` etc. but it is not required to install bower on your local machines since `npm` installs `bower` as a project dependency. Also `bower install` command is run automatically by `npm` as a post install script.

In addition we've created `npm scripts` to run `gulp tasks` and `gulp` is installed by npm as a project depency. So you don't have to install `gulp` or `gulp cli` manually on your working machine.

## Setting Up

- `clone` the repository
- `cd php-starter` to change the directory
- `npm install` to install npm dependencies
- `Composer install` to install composer dependencies

## Running the app

After you have installed all dependencies you can now run the app. Run `npm start` to start a local server using `gulp` which will watch your stylesheets and javascript files for changes, compile, minify, build and reload the browser automatically using `browser-sync` library.
The dev server will be opened in a new tab and usually on http://localhost:8000 and the Access URLs will be displayed on the terminal.

### server

```bash
# development
npm start

# production
npm run server:prod
```

# Configuration

The `gulp` tasks are defined inside the `gulpfile.babel.js` file found on the root of the application and you need to extend this file if you want to customize the application.

## Add third-party dependencies

### Composer

To integrate PHP libraries from composer, visite the packge repository website([https://packagist.org/](https://packagist.org/)) and follow the instructions. The command to install the package would generally look loke below. Replace the `$package` with the actual package you want to install.

```bash
composer require $package
```

Take a look at the bellow example:

Lets say that you want to integrate `facebook/graph-sdk` to your project. Just run the command bellow.

```bash
composer require facebook/graph-sdk
```

### Bower

To use any third party bower libraries other than the included `bootstrap`, `jquery` and `font-awesome` libs, locate the package you want in the [bower package repository](https://bower.io/search/) and use the following command to add it to your project.

```bash
bower install $package --save
```

Take a look at the bellow example:

Lets say that you want to add `moment.js` to your project. Just run the command bellow.

```bash
bower install moment --save
```

NOTE: Some packages won't work as expected and you might have to do overrides in the `bower.json` file.

# Styling

The `styles.scss` file inside the `sass` directory is the main stylesheet for the project and will be compiled and minified into an external `.css` and is embedded in the `index.html` file.
If you want to add your own stylesheet, we recommend that you place it under the `scss/partials` folder and import it in the `styles.scss` file.

For example if you want to include the styles for a slider:

1. Create a `_slider.scss` partial file in the `scss/partials` directory.
2. In `styles.scss` add `@import 'partials/slider.scss';`

# Testing

Execute the following command to run your unit tests. We use [codecov](https://codecov.io/) package to generate test coverage reports and the generated reports can be found in the `coverage` folder.

```bash
npm run test
```

# Linting

Execute the following commands to generate linting for styles and scripts.

```bash
# all
npm run lint

# javascript
npm run lint:scripts

# sass
npm run lint:styles
```

# Deployment

## Building the app

### For Development

Execute the following command to build your files in the development mode. A new folder called `dev` will be created and the artifacts will be saved there.

```bash
npm run build:dev
```

### For Production

Execute the following command to build your files in the production mode. A new folder called `dist` will be created and the artifacts will be saved there.

```bash
npm run build
```

## Deploy to Heroku

Deploy to [Heroku](https://www.heroku.com/) using this template with one single click. Heroku has an awesome free and it is extremely easy to use.

Click [here](https://php-starter-seed.herokuapp.com/) to see the deployed app.

<!-- Markdown snippet -->

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy)

# Built With

<a href="https://getcomposer.org/">
  <img src="./docs/readme-resources/technologies/composer.svg" alt="Composer" height="30" />
</a>&nbsp;&nbsp;
<a href="http://php.net/">
  <img src="./docs/readme-resources/technologies/php.png" alt="npm" height="25" />
</a>&nbsp;&nbsp;
<a href="https://www.npmjs.com/">
  <img src="./docs/readme-resources/technologies/npm.svg" alt="npm" height="20" />
</a>&nbsp;&nbsp;
<a href="https://bower.io/">
  <img src="./docs/readme-resources/technologies/bower.svg" alt="bower" height="30" />
</a>&nbsp;&nbsp;
<a href="https://sass-lang.com/">
  <img src="./docs/readme-resources/technologies/sass.svg" alt="sass" height="30" />
</a>&nbsp;&nbsp;
<a href="https://getbootstrap.com/docs/4.0/getting-started/introduction/">
  <img  src="./docs/readme-resources/technologies/boostrap.svg" alt="bootstrap" height="30" />
</a>&nbsp;&nbsp;
<a href="https://gulpjs.com/">
  <img style="display:inline-block;margin: 5px 10px" src="./docs/readme-resources/technologies/gulp.svg" alt="gulp" height="30" />
</a>

# Changelog

Please refer [CHANGELOG.md](CHANGELOG.md) to learn about the latest improvements, breaking changes and bug fixes.

# Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for contributing guidelines and to learn about our code of conduct.

# Contributors

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
| [<img src="https://avatars3.githubusercontent.com/u/25959096?v=4" width="80px;"/><br /><sub><b>Brion Mario</b></sub>](http://www.brionmario.com/)<br />[💻](https://github.com/brionmario/php-starter/commits?author=brionmario "Code") [📖](https://github.com/brionmario/php-starter/commits?author=brionmario "Documentation") [🐛](https://github.com/brionmario/php-starter/issues?q=author%3Abrionmario "Bug reports") [⚠️](https://github.com/brionmario/php-starter/commits?author=brionmario "Tests") |
| :---: |

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/kentcdodds/all-contributors) specification. Contributions of any kind welcome!

# License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details.


CHANGELOG.md
<a name="1.0.0-alpha.0"></a>
# 1.0.0-alpha.0 (2019-07-12)





src/composer-demo.php
<?php
  // require the Faker autoloader
  require_once './vendor/autoload.php';
?>

<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="msapplication-TileColor" content="#ffffff">
  <meta name="theme-color" content="#ffffff">
  <meta name="author" content="Brion Silva">
  <title>PHP Starter by Aparecium Labs</title>
  <!-- Favicon Package -->
  <link rel="apple-touch-icon" sizes="180x180" href="./assets/icons/favicon_package/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="./assets/icons/favicon_package/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="./assets/icons/favicon_package/favicon-16x16.png">
  <link rel="manifest" href="./assets/icons/favicon_package/site.webmanifest">
  <link rel="mask-icon" href="./assets/icons/favicon_package/safari-pinned-tab.svg" color="#5bbad5">

  <!-- inject:css -->
  <!-- endinject -->

  <!-- inject:js -->
  <!-- endinject -->
</head>

<body>
  <?php require './components/header.php'?><!-- header -->
  <div class="main-area">
    <div class="container">
      <div class="row">
        <div class="composer-demo text-center mt-5">
          <div class="heading">
            <img class="composer-logo mb-4" src="assets/img/composer.svg" height="300"/>
            <h6 class="text-dark">This starter project incorporates the power of composer. Please visit <a href="https://packagist.org/" target="_blank">https://packagist.org/</a> to browse packages and simply add them to to the project.</h6>
            <p>Following is a demonstration of the already included <span class="pre">faker</span> library. Press the generate button to generate a random word.</p>
          </div>
          <div class="faker-demo-container">
            <h4 id="fake-name" class="fake-name text-dark mt-4 mb-4">Fake Name</h4>
            <button id="generate-btn" class="btn btn-primary btn-sm">Generate</button>
          </div>
        </div>
      </div><!-- /row -->
    </div><!-- /container -->
  </div><!-- /main-area -->
  <?php require './components/footer.php'?><!--footer -->
  <script language="JavaScript" type="text/javascript">
    $(document).ready(function() {
      $("button").click(function() {
        <?php
        $faker = Faker\Factory::create();
        echo 'document.getElementById("fake-name").innerHTML = "'. $faker->name .'";';
        ?>
      });
    });
  </script>
</body>
</html>


src/components/footer.php
<div class="footer">
  <ul class="footer-icons">
    <li>
      <a href="https://www.facebook.com/apareciumlabs/" target="_blank"
        ><i class="fa fa-facebook"></i
      ></a>
    </li>
    <li>
      <a href="https://twitter.com/apareciumlabs" target="_blank"
        ><i class="fa fa-twitter"></i
      ></a>
    </li>
    <li>
      <a href="https://github.com/apareciumlabs" target="_blank"
        ><i class="fa fa-github"></i
      ></a>
    </li>
  </ul>
  <div class="copyright">
    <p>
      &copy; Copyright
      <script language="JavaScript" type="text/javascript">
        document.write(getCurrentYear());
      </script>
    </p>
    <img
      src="assets/img/logo_aparecium_fox_only.svg"
      width="20px"
      height="20px"
    />
    <p><a class="link" href="http://apareciumlabs.com" target="_blank">Aparecium Labs.</a> All rights reserved.</p>
  </div>
</div>


gulpfile.babel.js
import gulp from 'gulp';
import babel from 'gulp-babel';
import concat from 'gulp-concat';
import inject from 'gulp-inject';
import gutil from 'gulp-util';
import imagemin from 'gulp-imagemin';
import autoprefixer from 'gulp-autoprefixer';
import del from 'del';
import sass from 'gulp-sass';
import header from 'gulp-header';
import cleanCSS from 'gulp-clean-css';
import rename from 'gulp-rename';
import sassLint from 'gulp-sass-lint';
import eslint from 'gulp-eslint';
import pkg from './package.json';
import browserSync from 'browser-sync';
import uglify from 'gulp-uglify';
import sourcemaps from 'gulp-sourcemaps';
import wiredep from 'wiredep';
import urlAdjuster from 'gulp-css-url-adjuster';
import bowerlibs from 'main-bower-files';
import phpConnect from 'gulp-connect-php';

/**
 * Browser Support declaration
 * @type {string[]}
 */
const AUTO_PREFIX_BROWSERS = [
  'ie >= 10',
  'ie_mob >= 10',
  'ff >= 30',
  'chrome >= 34',
  'safari >= 7',
  'opera >= 23',
  'ios >= 7',
  'android >= 4.4',
  'bb >= 10'
];

/**
 * Banner for the compiled CSS files
 * @type {string}
 */
const banner = [
  '/*! ========================================================================= *\n',
  ' *  <%= pkg.title %> \n',
  ' *  ========================================================================= *\n',
  ' *  Version: <%= pkg.version %>\n',
  ' *  Author: <%= pkg.author %>\n',
  ' *  ========================================================================= *\n',
  ' *  GitHub Repo: <%= pkg.repository.url %>\n',
  ' *  ========================================================================= *\n',
  ' *  Copyright Ⓒ ' + new Date().getFullYear(),
  '\n',
  ' *  Licensed under <%= pkg.license %>\n',
  ' *  ========================================================================= *\n',
  ' */\n\n',
  ''
].join('');

const CONFIG = {
  filenames: {
    dev: {
      scripts: 'main.js',
      bowerJS: 'vendor.js',
      bowerCSS: 'vendor.css',
      styles: 'styles.css'
    },
    prod: {
      scripts: 'main.bundle.js',
      bowerJS: 'vendor.bundle.js',
      bowerCSS: 'vendor.bundle.css',
      styles: 'styles.bundle.css'
    }
  },
  paths: {
    dev: {
      root: 'dev',
      scripts: 'dev/js',
      styles: 'dev/css',
      pages: 'dev/pages',
      components: 'dev/components',
      assets: 'dev/assets',
      fonts: 'dev/assets/fonts',
      composer: 'dev/vendor'
    },
    src: {
      root: 'src',
      scripts: 'src/scripts',
      styles: 'src/sass',
      pages: 'src/pages',
      components: 'src/components',
      index: 'src/index.php',
      assets: 'src/assets',
      fonts: 'src/assets/fonts',
      bower: 'src/bower_components',
      composer: 'src/vendor'
    },
    prod: {
      root: 'dist',
      scripts: 'dist/js',
      styles: 'dist/css',
      pages: 'dist/pages',
      assets: 'dist/assets',
      fonts: 'dist/assets/fonts',
      components: 'dist/components',
      composer: 'dist/vendor'
    }
  },
  settings: {
    dev: {
      proxy: '127.0.0.1:8010',
      tunnel: 8010,
      port: 8000
    },
    prod: {
      proxy: '127.0.0.1:8010',
      tunnel: 8010,
      port: 3000
    }
  }
};

gulp.task('scripts:build', () => {
  return gulp
    .src([`${CONFIG.paths.src.scripts}/**/*.js`, `!${CONFIG.paths.src.scripts}/**/*.test.js`])
    .pipe(gutil.env.env === 'production' ? gutil.noop() : sourcemaps.init())
    .pipe(
      babel({
        presets: ['@babel/env']
      })
    )
    .pipe(gutil.env.env === 'production' ? gutil.noop() : sourcemaps.write())
    .pipe(gutil.env.env === 'production' ? concat(CONFIG.filenames.prod.scripts) : gutil.noop())
    .pipe(gutil.env.env === 'production' ? uglify() : gutil.noop())
    .pipe(
      gutil.env.env === 'production'
        ? rename({
          suffix: '.min'
        })
        : gutil.noop()
    )
    .pipe(
      gutil.env.env === 'production'
        ? header(banner, {
          pkg: pkg
        })
        : gutil.noop()
    )
    .pipe(
      gutil.env.env === 'production'
        ? gulp.dest(`${CONFIG.paths.prod.scripts}`)
        : gulp.dest(`${CONFIG.paths.dev.scripts}`)
    );
});

gulp.task('scripts:lint', () => {
  return gulp
    .src(`${CONFIG.paths.src.scripts}/**/*.js`)
    .pipe(eslint())
    .pipe(eslint.format());
});

gulp.task('styles:build', () => {
  return gulp
    .src(`${CONFIG.paths.src.styles}/**/*.s+(a|c)ss`)
    .pipe(
      sass({
        outputStyle: 'nested',
        precision: 10,
        includePaths: ['.'],
        onError: console.error.bind(console, 'Sass error:')
      })
    )
    .pipe(autoprefixer({ browsers: AUTO_PREFIX_BROWSERS }))
    .pipe(
      urlAdjuster({
        prependRelative: '../'
      })
    )
    .pipe(
      gutil.env.env === 'production'
        ? concat(CONFIG.filenames.prod.styles)
        : concat(CONFIG.filenames.dev.styles)
    )
    .pipe(
      gutil.env.env === 'production'
        ? header(banner, {
          pkg: pkg
        })
        : gutil.noop()
    )
    .pipe(gutil.env.env === 'production' ? cleanCSS() : gutil.noop())
    .pipe(
      gutil.env.env === 'production'
        ? rename({
          suffix: '.min'
        })
        : gutil.noop()
    )
    .pipe(
      gutil.env.env === 'production'
        ? gulp.dest(CONFIG.paths.prod.styles)
        : gulp.dest(CONFIG.paths.dev.styles)
    )
    .pipe(browserSync.stream());
});

gulp.task('styles:lint', () => {
  return gulp
    .src(`${CONFIG.paths.src.root}/**/*.s+(a|c)ss`)
    .pipe(
      sassLint({
        options: {
          formatter: 'stylish',
          'merge-default-rules': false
        },
        files: { ignore: 'node_modules/!**!/!*.s+(a|c)ss' },
        rules: {
          'no-ids': 1,
          'no-mergeable-selectors': 0
        },
        configFile: '.sass-lint.yml'
      })
    )
    .pipe(sassLint.format());
});

gulp.task('clean', () => {
  if (gutil.env.env === 'production') {
    return del(CONFIG.paths.prod.root);
  }
  if (gutil.env.env === 'development') {
    return del(CONFIG.paths.dev.root);
  }
  if (gutil.env.env === 'all') {
    return del([CONFIG.paths.dev.root, CONFIG.paths.prod.root]);
  }
  return del([CONFIG.paths.dev.root, CONFIG.paths.prod.root]);
});

gulp.task('move:assets', () => {
  return gulp
    .src([`${CONFIG.paths.src.assets}/**/*`], { base: CONFIG.paths.src.root })
    .pipe(gutil.env.env === 'production' ? imagemin() : gutil.noop())
    .pipe(
      gutil.env.env === 'production'
        ? gulp.dest(CONFIG.paths.prod.root)
        : gulp.dest(CONFIG.paths.dev.root)
    );
});

gulp.task('move:bower:fonts', () => {
  return gulp
    .src(bowerlibs('**/*.{eot,svg,ttf,woff,woff2}'))
    .pipe(
      gutil.env.env === 'production'
        ? gulp.dest(CONFIG.paths.prod.fonts)
        : gulp.dest(CONFIG.paths.dev.fonts)
    );
});

gulp.task('move:composer:deps', () => {
  return gulp
    .src([`${CONFIG.paths.src.composer}/**/*`], { base: CONFIG.paths.src.root })
    .pipe(
      gutil.env.env === 'production'
        ? gulp.dest(CONFIG.paths.prod.composer)
        : gulp.dest(CONFIG.paths.dev.composer)
    );
});

gulp.task('bundle:bower', () => {
  let target = gulp.src(
    [
      CONFIG.paths.src.index,
      `${CONFIG.paths.src.components}/**/*.{html|jade|php}`,
      `${CONFIG.paths.src.pages}/**/*.{html|jade|php}`
    ],
    { base: CONFIG.paths.src.root }
  );

  let js = gulp.src(wiredep().js);
  let css = gulp.src(wiredep().css);

  return target
    .pipe(
      inject(
        js
          .pipe(
            concat(
              gutil.env.env === 'production'
                ? CONFIG.filenames.prod.bowerJS
                : CONFIG.filenames.dev.bowerJS
            )
          )
          .pipe(
            gutil.env.env === 'production'
              ? gulp.dest(CONFIG.paths.prod.scripts)
              : gulp.dest(CONFIG.paths.dev.scripts)
          )
      )
    )
    .pipe(
      inject(
        css
          .pipe(
            concat(
              gutil.env.env === 'production'
                ? CONFIG.filenames.prod.bowerCSS
                : CONFIG.filenames.dev.bowerCSS
            )
          )
          .pipe(
            urlAdjuster({
              replace: ['../fonts', '../assets/fonts']
            })
          )
          .pipe(
            gutil.env.env === 'production'
              ? gulp.dest(CONFIG.paths.prod.styles)
              : gulp.dest(CONFIG.paths.dev.styles)
          )
      )
    );
});

gulp.task('inject', () => {
  let target = gulp.src(
    [
      `${CONFIG.paths.src.root}/**/*.html`,
      `${CONFIG.paths.src.root}/**/*.jade`,
      `${CONFIG.paths.src.root}/**/*.php`,
      `!${CONFIG.paths.src.bower}/**/*`
    ],
    { base: CONFIG.paths.src.root }
  );

  let devSources = gulp.src(
    [
      `${CONFIG.paths.dev.scripts}/**/*.js`,
      `${CONFIG.paths.dev.styles}/**/*.css`,
      `!${CONFIG.paths.dev.scripts}/**/*.test.js`
    ],
    {
      read: false
    }
  );

  let devInjectionOptions = {
    ignorePath: CONFIG.paths.dev.root,
    addRootSlash: false
  };

  let prodSources = gulp.src(
    [
      `${CONFIG.paths.prod.scripts}/**/*.js`,
      `${CONFIG.paths.prod.styles}/**/*.css`,
      `!${CONFIG.paths.prod.scripts}/**/*.test.js`
    ],
    { read: false }
  );

  let prodInjectionOptions = {
    ignorePath: CONFIG.paths.prod.root,
    addRootSlash: false
  };

  return target
    .pipe(
      gutil.env.env === 'production'
        ? inject(prodSources, prodInjectionOptions)
        : inject(devSources, devInjectionOptions)
    )
    .pipe(
      gutil.env.env === 'production'
        ? gulp.dest(CONFIG.paths.prod.root)
        : gulp.dest(CONFIG.paths.dev.root)
    );
});

gulp.task(
  'build',
  gulp.series(
    'move:composer:deps',
    'move:assets',
    'move:bower:fonts',
    gulp.parallel('scripts:build', 'styles:build'),
    'bundle:bower',
    'inject'
  ),
  callback => {
    callback();
  }
);

gulp.task('php', callback => {
  let root = CONFIG.paths.dev.root;
  let port = CONFIG.settings.dev.tunnel;

  if (gutil.env.env === 'production') {
    root = CONFIG.paths.prod.root;
    port = CONFIG.settings.prod.tunnel;
  }

  phpConnect.server({
    base: root,
    port: port,
    keepalive: true
  });
  callback();
});

gulp.task('browserSync', callback => {
  let proxy = CONFIG.settings.dev.proxy;
  let port = CONFIG.settings.dev.port;

  if (gutil.env.env === 'production') {
    proxy = CONFIG.settings.prod.proxy;
    port = CONFIG.settings.prod.port;
  }

  browserSync.init({
    proxy: proxy,
    port: port,
    open: true,
    notify: false
  });
  callback();
});

gulp.task('watch:styles', () => {
  gulp.watch(`${CONFIG.paths.src.styles}/**/*.s+(a|c)ss`, gulp.series('styles:build'));
});

gulp.task('watch:scripts', () => {
  gulp
    .watch(`${CONFIG.paths.src.scripts}/**/*.js`, gulp.series('scripts:build'))
    .on('change', browserSync.reload);
});

gulp.task('watch:pages', () => {
  gulp
    .watch(
      [
        `${CONFIG.paths.src.root}/**/*.php`,
        `${CONFIG.paths.src.root}/**/*.html`,
        `${CONFIG.paths.src.root}/**/*.jade`
      ],
      gulp.series('inject')
    )
    .on('change', browserSync.reload);
});

gulp.task('watch:assets', () => {
  gulp
    .watch(`${CONFIG.paths.src.assets}/**/*`, gulp.series('move:assets'))
    .on('change', browserSync.reload);
});

gulp.task('watch', gulp.parallel('watch:styles', 'watch:scripts', 'watch:pages', 'watch:assets'));

gulp.task('serve', gulp.series('browserSync', 'php', 'watch'));
